sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypes.scala

```diff
@@ -79,3 +78,42 @@ case class CreateStruct(children: Seq[Expression]) extends Expression {
     InternalRow(children.map(_.eval(input)): _*)
   }
 }
+
+/**
+ * Creates a struct with the given field names and values
+ *
+ * @param children Seq(name1, val1, name2, val2, ...)
+ */
+case class CreateNamedStruct(children: Seq[Expression]) extends Expression {
+  assert(children.size % 2 == 0, "NamedStruct expects an even number of arguments.")
+
+  private val nameExprs = children.zipWithIndex.filter(_._2 % 2 == 0).map(_._1)
+  private val valExprs = children.zipWithIndex.filter(_._2 % 2 == 1).map(_._1)
+
+  private lazy val names = nameExprs.map { case name =>
+    name match {
+      case NonNullLiteral(str, StringType) =>
+        str.asInstanceOf[UTF8String].toString
+      case _ =>
+        throw new IllegalArgumentException("Expressions of odd index should be" +
+          s" Literal(_, StringType), get ${name.dataType} instead")
+    }
+  }
```

Comment:
With `checkInputTypes`, we can ensure when we reach here, `nameExprs` is of correct type and we can simplify the code here as `nameExprs.map(_.asInstanceOf[Literal].value.toString)`